### PR TITLE
#8766: Timeline plugin configuration of snap and mode on startup

### PR DESCRIPTION
**Pull request by dsuren1**
## Description
This PR allow configuration of new options in Timeline plugin
- `initialMode` - `single|range`
- `initialSnap` - `now|fullRange`
- `resetButton` - boolean

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Enhancement

## Issue

**What is the current behavior?**
- https://github.com/geosolutions-it/MapStore2/issues/8766

**What is the new behavior?**
Based on the configuration the timeline plugin is initialized and time is set on the timeline slider
Kindly refer the issue for scenarios

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/reducers/dimension.js [hunk 1]
```diff
@@ -58,6 +59,13 @@ export default (state = {}, action) => {
     case RESET_CONTROLS: {
         return set('data', undefined, set('currentTime', undefined, set('offsetTime', undefined, state)));
     }
+    case MAP_CONFIG_LOADED: {
+        const config = action?.config || {};
+        return {...state,
+            currentTime: config?.dimensionData?.currentTime,
+            offsetTime: config?.dimensionData?.offsetTime
+        };
+    }
```
Comment (offtherailz) [inline web/client/reducers/dimension.js:68]: I should keep the epic in the middle if possible, or we can not control or avoid this operation with a flag. 
This is technically an operation that the timeline plugin does to initialize. Not a standard behavior of the dimensions state.

**Review by offtherailz — CHANGES_REQUESTED** [review]
For the moment follow the suggestions I asked. I wasn't able to test it yet (I need to set-up some sample time data to do that). 
Anyway you can start working on it.

File: web/client/epics/dimension.js [hunk 2]
```diff
@@ -102,16 +100,13 @@ export const updateLayerDimensionDataOnMapLoad = (action$, {getState = () => {}}
...
-        const offsetTime = config.dimensionData?.offsetTime;
 
         return (currentTime ?
             Observable.of(
                 // restore states of timeline and dimension from map config
                 ...(selectedLayer ? [
                     initializeSelectLayer(selectedLayer)
-                ] : []),
-                setCurrentTime(currentTime),
-                setCurrentOffset(offsetTime)
```
[the remark this answers is not in the document]
Comment (dsuren1) [reply web/client/epics/dimension.js:114]: Kindly correct me if I'm wrong, instead we should have an epic in timeline to dispatch `initializeSelectedLayer` on `MAP_CONFIG_LOADED` and remove it from here(dimension) while retaining the dispatch action of dimension data and remove the `MAP_CONFIG_LOADED` from the dimension reducer based on this [comment ](https://github.com/geosolutions-it/MapStore2/pull/8788#discussion_r1018197891) ?

[the remark this answers is not in the document]
Comment (offtherailz) [reply web/client/epics/dimension.js:114]: Yes. Also  `autoselect` and `setEndValuesSupport` should be delegated to timeline. 

In general `dimension` should not depend from timeline. 

I see `autoselect` and `setEndValuesSupport` are triggered after `updateLayerDimensionData`. 
So maybe we can delegate all these to the timeline:
- `initializeSelectedLayer` on `MAP_CONFIG_LOADED`
- `autoselect` `setEndValuesSupport` after `UPDATE_LAYER_DIMENSION_DATA`

I think this helps to isolate the purposes of the various parts:
- dimension takes care of keeping updated the dimension data of the layers and the map and time
- timeline /playback update the current time, range, downloads the proper dimension data (invoking dimension) accordingly with the current viewport of the timeline and so on. 



Moreover looking at the code, I see that `setEndValuesSupport` is triggered only after `ADD_LAYER` but not on `MAP_CONFIG_LOADED`, that looks incorrect to me. (I'm afraid that saving a map with layers with start-end values and then reloading may have different results ). Could you please check?

**Comment by offtherailz** [comment]
Hi @dsuren1
I asked to change the icon to clear up the meanings. In the meanwhile I reviewed this PR: https://github.com/geosolutions-it/MapStore2/pull/8793

This makes me think about a way to clarify and unify similar concepts graphically, so I discussed with @allyoucanmap about it. 

In both the application we have an functionality `filter data by viewport` that should be have the same icon for both the feature grid and the timeline, becuase they have similar meaning (filter the content of the plugin on the current map viewport`.

For this reason, asking to @tdipisa first, we'd like to create a new icon "Filter by viewport" to use for both the PRs.

So I should wait to update the documentation to use the new icon.
The new icon needs to similar to this 
zoom-to
But with the filter on center instead of the magifier
filter

File: web/client/epics/dimension.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (dsuren1) [reply web/client/epics/dimension.js:114]: @offtherailz 
> Moreover looking at the code, I see that setEndValuesSupport is triggered only after ADD_LAYER but not on MAP_CONFIG_LOADED, that looks incorrect to me. (I'm afraid that saving a map with layers with start-end values and then reloading may have different results ). Could you please check?

Looks like `endValuesSupport` & `snapRadioButtonEnabled` are being set directly in the timeline reducer on `MAP_CONFIG_LOADED`

Also I did stumble upon [this](https://github.com/geosolutions-it/MapStore2/blob/master/web/client/reducers/timeline.js#L107) where the snapType is always set to 'start' on `SELECT_LAYER`. Could you please clarify if this is intended?

**Comment by dsuren1** [comment]
@offtherailz 
Except for new icon the rest have been updated as per suggestion

**Comment by dsuren1** [comment]
@offtherailz 
Kindly clarify the following. Thanks!
> * [ ]  I tried to click reset with `"initialMode": "single", "initialSnap": "now"` but it didn't worked. Please  make sure that it works. A second time I have a strange zoom behavour:

Could you please share me the test data (time layer) used, as I'm unable to reproduce this behavior with the time layers present in _gs-stable_ (`meteroite` and `storm-obs`)

> * [ ] The reset button should show "set range to include all data" with "<->" icon only when range mode is active.

_Should the reset button be enabled only in these two scenarios i.e. `single-now` and `range -fullRange`? What about `range-now`, is this an invalid combo?_

> * [ ]  In single mode, the reset button should show a clock icon and the text "set time to now" if snapping is disabled, "set time to nearest value to now" if snapping is enabled

Since snapping and layer selection is directly related (i.e. with autoselect, enabling one will select the other). With current implementation, hitting reset would select the layer when not selected which in turn trigger epic to `snap to nearest now`. 
With this requirement, in this case(`single-now` - layer unselected/snapping is disabled), should the layer be left _unselected_?

**Comment by offtherailz** [comment]
> Could you please share me the test data (time layer) used, as I'm unable to reproduce this behavior with the time layers present in _gs-stable_ (`meteroite` and `storm-obs`)

This is my steps sequence:
- Configure timeline as following: 
```json
{
              "name": "Timeline",
              "cfg": {
                "expandLimit": 20,
                "resetButton": true,
                "initialMode": "single",
                "initialSnap": "now"
              }
            },
```
- create a new map
- add storm_obs from catalog
- open the timeline and click on a time different from now
- click on reset button (nothing happens)

https://user-images.githubusercontent.com/1279510/201619768-bfab2b3b-4e66-4ec4-99a9-55c2f7986e68.mp4

Anyway if you are going to change the logic, probably it will not make difference. 

> Should the reset button be enabled only in these two scenarios i.e. single-now and range -fullRange? What about range-now, is this an invalid combo?

The button at this point do not depend from initial state anymore. It depends from the current mode only. Resets to full range if we are in range mode, resets to now if we are in single time mode. I should rely only on this now, in order to make things more simple and independent. 

> Since snapping and layer selection is directly related (i.e. with autoselect, enabling one will select the other). With current implementation, hitting reset would select the layer when not selected which in turn trigger epic to snap to nearest now.
With this requirement, in this case(single-now - layer unselected/snapping is disabled), should the layer be left unselected?

If so, we should correct it. The behavior of reset should be the same of clicking on now on the timeline. The click event cause a sequence that sets time to now if the layer is not selected, or to the nearest value of now if not. Find out where this happens and reuse the same flow, if possible.

**Comment by offtherailz** [comment]
@dsuren1 https://github.com/geosolutions-it/MapStore2/pull/8807 the icon has been merged. Please align to master to get the right icon

**Comment by offtherailz** [comment]
@dsuren1 I updated the JSDoc by testing it with `npm start` + `npm run jsdoc:test`.  
Please take a look if it makes sense to you too.

**Comment by dsuren1** [comment]
@offtherailz Thanks. It looks fine.

**Comment by offtherailz** [comment]
Please @ElenaGallo test this on DEV. 
You can do it by creating a context and modifying the properties provided in the doc checking that:
- The reset button does what it need, and it is visible when enabled
- When add (the first) time layer, the timeline sets to range / single mode if configured
- When add (the first) time layer, the timeline sets to include all data, if `initialSnap`= `fullRange` is configured

**Pull request merged**